util/gem5_bridge/src/gem5_bridge.h

```diff
@@ -0,0 +1,46 @@
+/* SPDX-License-Identifier: BSD AND GPL-2.0+ */
+/*
+ * Copyright (c) 2024 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __GEM5_BRIDGE_H__
+#define __GEM5_BRIDGE_H__
+
+/* Physical address for base of gem5ops MMIO range */
+#ifndef GEM5OPS_BASE
+#define GEM5OPS_BASE 0xffff0000 /* default, also used for x86 */
```

Comment:
This really ought to be discovered from the device tree or an ACPI table to make it a bit more robust. There is a good example here: https://elixir.bootlin.com/linux/v6.10.8/source/drivers/virtio/virtio_mmio.c

Given that gem5's ACPI implementation is very basic, it might be better and easier to just emulate a PCI device instead. That would be reasonably easy to implement on the gem5 side and would just work.

Reply:
I agree that using the devicetree/acpi would be cool, but I think that can be a second step.

Reply:
This has been changed as well. The driver now accepts arguments for these upon insertion, which supports both default values as defined by the disk image init script (where the module is inserted) as well as kernel argument overloads that come from the sim config, if needed.

Reply:
Can you please explain why this request would be a blocker to getting this initial implementation merged? From my perspective, it's not an insignificant undertaking to learn how ACPI and DT works. This change currently adds a nice new feature to gem5, and it would be shame to block it due to it not being "nice enough." If I'm mistaken, and there is something that is truly blocking here, please explain in more detail.

I don't disagree at all that it would be nice to have device tree/ACPI support. But it seems to me that this can be done after the initial change is merged.

Reply:
This [platform API](https://docs.kernel.org/driver-api/driver-model/platform.html) is designed for simple, typically memory mapped, devices and provides a standardised way of getting to the associated MMIOs and IRQs. If you use this API, you get more convenient  managed (see the [devres manual](https://docs.kernel.org/driver-api/driver-model/devres.html)) IO ranges and interrupts, this simplifies initialisation and clean up. The other big benefit is that Device Tree can both be supported by simply adding a table with the compatible strings. I.e., there is no need to actually understand any device tree APIs in the basic case, things will just work automatically since MMIO and interrupt assignment is standardised.

Contrary to my previous comment, the only drawback in this case seems to be that the non-DT/ACPI  will need custom parameter parsing. The [VirtIO MMIO](https://elixir.bootlin.com/linux/v6.10.8/source/drivers/virtio/virtio_mmio.c) has a very good example of that though.

In practice, I believe that this rewrite would practically be needed to support device tree anyway, so we might as well do it now.